ts/private/ts_project_worker.js

```diff
@@ -463,111 +652,132 @@ function createProgram(args, initialInputs) {
         if (args.join(' ') != newArgs.join(' ')) {
-            const {options} = ts.parseCommandLine(newArgs);
+            debug(`arguments have changed.`);
+            debug(`  current: ${newArgs.join(" ")}`);
+            debug(`  previous: ${args.join(" ")}`);
+
+            const { options } = ts.parseCommandLine(newArgs);
             for (const key in compilerOptions) {
-                delete compilerOptions[key];
+                if (!(key in options)) {
+                    delete compilerOptions[key];
+                }
             }
             for (const key in options) {
                 compilerOptions[key] = options[key];
             }
-            enablePerformanceAndTracingIfNeeded();
+
+            disableStatisticsAndTracing();
+            enableStatisticsAndTracing();
             updateOutputs();
+            applySyntheticOutPaths();
             // invalidating tsconfig will cause parseConfigFile to be invoked
-            filesystemTree.update(tsconfig);
+            filesystem.update(tsconfig);
             args = newArgs;
         }
     }
 
-    function debuglog(message) {
-        // TODO: https://github.com/aspect-build/rules_ts/issues/189
-        compilerOptions.extendedDiagnostics && host.trace?.(message)
-    }
-
     function readFile(filePath, encoding) {
-        const relative = path.relative(execRoot, filePath);
-        // external lib are transitive sources thus not listed in the inputs map reported by bazel.
-        if (!filesystemTree.fileExists(path.relative(execRoot, filePath)) && !isExternalLib(filePath) && !outputs.has(relative)) {
-            throw new Error(`tsc tried to read file (${filePath}) that wasn't an input to it.`);
-        }
-        return ts.sys.readFile(filePath, encoding);
-    }
+        filePath = path.resolve(sys.getCurrentDirectory(), filePath)
 
-    function directoryExists(directoryPath) {
-        if (!directoryPath.startsWith(bin)) {
-            return false;
+        //external lib are transitive sources thus not listed in the inputs map reported by bazel.
+        if (!filesystem.fileExists(filePath) && !isExternalLib(filePath) && !outputs.has(filePath)) {
+            output.write(`tsc tried to read file (${filePath}) that wasn't an input to it.` + "\n");
+            //throw new Error(`tsc tried to read file (${filePath}) that wasn't an input to it.`);
         }
-        return filesystemTree.directoryExists(path.relative(execRoot, directoryPath));
-    }
 
-    function getDirectories(directoryPath) {
-        return filesystemTree.getDirectories(path.relative(execRoot, directoryPath));
+        return ts.sys.readFile(path.join(execroot, filePath), encoding);
     }
 
-    function fileExists(filePath) {
-        return filesystemTree.fileExists(path.relative(execRoot, filePath));
+    function createDirectory(p) {
+        // TODO: cleanup
+        debug("createDirectory", p);
+        ts.sys.createDirectory(p.replace("__synthetic__outdir__", "."));
```

Comment:
nit: put `__synthetic__outdir__` in a const since it is used in multiple places. You could drop the `__` in the middle as well and make it just `__synthetic_outdir__`.... or for clarity in error messages (if it is going to show up there) make it `__rules_ts_worker__`